aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -799,4 +800,241 @@ Tensor sspaddmm(const Tensor& self, const Tensor& mat1, const Tensor& mat2,
+  }
+
+  LongTensor indices = input._indices();
+  Tensor values = input._values();
+  IntList sizes = input.sizes();
+  const int64_t sparse_dim = input.sparse_dim();
+  const int64_t dense_dim = input.dense_dim();
+
+  int64_t sparse_dims_to_sum_size = 0;
+  auto dims_to_keep_v = std::vector<int64_t>();
+  Tensor sparse_dims_to_keep_t = at::zeros({sparse_dim}, indices.options().dtype(at::kByte));
+  auto dense_dims_to_sum_v = std::vector<int64_t>();
+  for (int64_t d = 0; d < input_dim; d++) {
+    if (dims_to_sum_b[d]) {
+      if (d < sparse_dim) sparse_dims_to_sum_size ++;
+      else dense_dims_to_sum_v.emplace_back(d + 1 - sparse_dim);
+    }
+    else {
+      dims_to_keep_v.emplace_back(d);
+      if (d < sparse_dim) sparse_dims_to_keep_t[d].fill_(1);
+    }
+  }
+
+  const bool sum_all_sparse_dim = (sparse_dim == sparse_dims_to_sum_size);
+
+  // new indices
+  LongTensor new_indices;
+  if (keepdim) {
+    new_indices = indices.clone();
+    if (sparse_dims_to_sum_size > 0) {
+      for (auto d : dims_to_sum_v) {
+        if (d < sparse_dim) new_indices[d].fill_(0);
+      }
+    }
+  }
+  else {
+    if (sum_all_sparse_dim) {
+      new_indices = at::zeros({1, input._nnz()}, indices.options());
+    }
+    else if (sparse_dims_to_sum_size > 0) {
+      new_indices = indices.index_select(0, at::arange(sparse_dim, indices.options()).masked_select(sparse_dims_to_keep_t));
+    }
+    else {
+      new_indices = indices.clone();
+    }
+  }
+
+  // new size
+  std::vector<int64_t> new_sizes;
+  if (keepdim) {
+    new_sizes = sizes.vec();
+    for (auto d : dims_to_sum) new_sizes[d] = 1;
+  }
+  else {
+    for (auto d : dims_to_keep_v) new_sizes.emplace_back(sizes[d]);
+    if (sum_all_sparse_dim) new_sizes.emplace(new_sizes.begin(), 1);
+  }
+
+  // new values
+  Tensor new_values = values.clone();
+  if (dense_dims_to_sum_v.size() > 0) {
+    new_values = new_values.sum(dense_dims_to_sum_v);
+
+    if (keepdim) {
+      auto new_values_size = values.sizes().vec();
+      for (auto d : dense_dims_to_sum_v) new_values_size[d] = 1;
+      new_values = new_values.reshape(new_values_size);
+    }
+  }
+
+  SparseTensor new_sparse = at::sparse_coo_tensor(new_indices.contiguous(), new_values.contiguous(), new_sizes, input.options());
+  new_sparse = new_sparse.coalesce();
+
+  if (sum_all_sparse_dim) {
+    // return a dense tensor if sum over all sparse dims
+    if (keepdim) return new_sparse.to_dense();
+    else return new_sparse._values().squeeze();
+  }
+  return new_sparse;
+}
+
+Tensor _sparse_sum(const SparseTensor& input, IntList dims_to_sum, bool keepdim, ScalarType dtype) {
+  return at::_sparse_sum(input.to(dtype), dims_to_sum, keepdim);
+}
+
+// --------------------------------------------------------------------
+// NOTE [ sparse.sum() backward ]
+//
+// When sum over sparse_dim, backward needs to scatter gradients to nnz locations.
+// To do that, indices of grad and input need to be aligned over dims that are not summed.
```

Comment:
I found this sentence a little confusing, because I initially parsed this as "the indexes of the sparse tensor need to be aligned on nnz"; but, I think what you actually mean is, align the indices on *dimensions*. An example would be worth a million words here.

Reply:
Hmm... having looked at the code now, maybe you did mean nnz here.

Reply:
Sorry my prev NOTE was very difficult to read, let me try again with a simple example. I mainly want to note down how edge cases are handled in the backward, and to avoid surprises when using it.